Task4

lemonclouuds opened the pull request.

lemonclouuds committed:
Created cart page + serializable

lemonclouuds committed:
Implemented cart update

lemonclouuds committed:
Implemented deleting cart item

lemonclouuds committed:
Implemented minicart

lemonclouuds committed:
Added "Add to cart" button to PLP

lemonclouuds committed:
Added tests

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -44,18 +44,60 @@ public synchronized Cart getCart(HttpServletRequest request) {
+        );
+
+        BigDecimal[] totalCost = {new BigDecimal(0)};
+        cart.getItems().forEach(item -> {
```
this calculations can be done using only Stream API

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -44,18 +44,60 @@ public synchronized Cart getCart(HttpServletRequest request) {
+    @Override
+    public Optional<CartItem> findCartItem(Cart cart, Long productId, int quantity) throws OutOfStockException {
+        Product product = productDao.getProduct(productId);
+        if (product.getStock() < quantity) {
```
stock check is not expected in method named findCartItem

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -44,18 +44,60 @@ public synchronized Cart getCart(HttpServletRequest request) {
+            recalculateCart(cart);
             return;
         }
+        CartItem cartItem = new CartItem(product, quantity);
```
updateCart method should only update existing cart item
for new cart item creation we have addProductToCart

Alexandra-ES commented on src/main/java/com/es/phoneshop/web/DeleteCartItemServlet.java:
```diff
@@ -0,0 +1,35 @@
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        String productId = request.getPathInfo().substring(1);
+        Cart cart = cartService.getCart(request);
+        cartService.deleteProductFromCart(cart, Long.valueOf(productId));
```
handle parsing exception

Alexandra-ES commented on src/main/webapp/WEB-INF/pages/cart.jsp:
```diff
@@ -0,0 +1,80 @@
+<tags:master pageTitle="Cart">
+<form method="post" action="${pageContext.servletContext.contextPath}/cart">
+  <p>
+    Cart: ${cart}, total quantity: ${cart.totalQuantity}
```
remove this line, it was only for demo purposes

Alexandra-ES commented on src/main/webapp/WEB-INF/pages/cart.jsp:
```diff
@@ -0,0 +1,80 @@
+        <td></td>
+        <td>Total cost</td>
+        <td>
+        <fmt:formatNumber value="${cart.totalCost}" type="currency" currencySymbol="${item.product.currency.symbol}"/>
```
currency symbol is not displayed

Alexandra-ES commented on src/main/webapp/WEB-INF/pages/productList.jsp:
```diff
@@ -42,6 +43,12 @@
                 <fmt:formatNumber value="${product.price}" type="currency" currencySymbol="${product.currency.symbol}"/>
             </a>
         </td>
+        <form method="post" action="${pageContext.servletContext.contextPath}/products/${product.id}">
```
not the best choice of url

Alexandra-ES commented on src/main/webapp/WEB-INF/tags/master.tag:
```diff
@@ -13,6 +13,8 @@
       PhoneShop
     </a>
+    <jsp include page="/cart/minicart"/>
+    Cart: ${cart.totalQuantity}
```
remove this line

Alexandra-ES commented on src/test/java/com/es/phoneshop/model/product/CartServiceTest.java:
```diff
@@ -6,14 +6,15 @@
     private ProductDao productDao;
     private CartService cartService;
     Currency usd = Currency.getInstance("USD");
-    int quantity;
+    int quantity = 5;
```
can be private and final

Alexandra-ES commented on src/test/java/com/es/phoneshop/model/product/CartServiceTest.java:
```diff
@@ -66,9 +65,48 @@ public void testFailedAddProductToCart() throws OutOfStockException {
+        cartService.addProductToCart(cart, product.getId(), quantity);
+
+        Optional<CartItem> result = cartService.findCartItem(cart, product.getId(), quantity);
+        if (result.isPresent()) {
```
you can make this check to be part of test assertions

Alexandra-ES commented on src/test/java/com/es/phoneshop/web/CartPageServletTest.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Test
+    public void testDoPost() throws ServletException, IOException {
```
[opt] could you write a test to check servlet behaviour in case of errors

Alexandra-ES commented on src/test/java/com/es/phoneshop/model/product/CartServiceTest.java:
```diff
@@ -66,9 +65,48 @@ public void testFailedAddProductToCart() throws OutOfStockException {
+    }
+
+    @Test
+    public void testUpdateCart() throws OutOfStockException {
```
[opt] add test cases with OutOfStockException

lemonclouuds committed:
Fixes
-fixed minicart
-fixed methods in DefaultCartService
-added tests for exceptional situations

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -33,17 +33,25 @@ public static CartService getInstance() {
     public synchronized void addProductToCart(Cart cart, Long productId, int quantity) throws OutOfStockException {
         Product product = productDao.getProduct(productId);
+        if (quantity <= 0) {
+            throw new OutOfStockException(product, quantity, product.getStock());
```
use IllegalArgumentException instead

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -59,16 +67,20 @@ public synchronized void addProductToCart(Cart cart, Long productId, int quantit
     public void updateCart(Cart cart, Long productId, int quantity) throws OutOfStockException {
         Product product = productDao.getProduct(productId);
+        if (quantity <= 0) {
+            throw new OutOfStockException(product, quantity, product.getStock());
```
use IllegalArgumentException instead

lemonclouuds committed:
Fixed minicart, added IllegalArgument

lemonclouuds committed:
Removed productId from redirect
